sdk/compiler/damlc/lib/DA/Cli/Damlc/Command/MultiIde/SdkInstall.hs

```diff
@@ -0,0 +1,254 @@
+        finishSdkInstallReporter miState ver
+      changeSdkStatus LSP.DsInfo (installingSdkIdeDiagnosticMessage ver) (SISInstalling installThread)
+    (SISAsking, _) -> disableSdk
+    (_, _) -> atomically $ putTMVar (misSdkInstallDatasVar miState) installDatas
+handleShowMessageResponse _ _ _ = pure ()
+
+-- try @SomeException that doesn't catch AsyncCancelled exceptions
+tryForwardAsync :: IO a -> IO (Either SomeException a)
+tryForwardAsync = tryJust @SomeException $ \case
+  (fromException -> Just AsyncCancelled) -> Nothing
+  e -> Just e
+
+handleSdkInstallCancelled :: MultiIdeState -> LSP.NotificationMessage 'LSP.CustomMethod -> IO ()
+handleSdkInstallCancelled miState notif = do
+  forM_ (fromJSON $ notif ^. LSP.params) $ \message -> do
+    let ver = sicSdkVersion message
+    installDatas <- atomically $ takeTMVar $ misSdkInstallDatasVar miState
+    let installData = getSdkInstallData ver installDatas
+    installDatas' <- case sidStatus installData of
+      SISInstalling thread -> do
+        logDebug miState $ "Killing install thread for " <> unresolvedReleaseVersionToString ver
+        cancel thread
+        updateSdkStatus miState installDatas ver LSP.DsError (missingSdkIdeDiagnosticMessage ver) SISDenied
+      _ -> pure installDatas
+    atomically $ putTMVar (misSdkInstallDatasVar miState) installDatas'
+
+handleInstallResult :: MultiIdeState -> UnresolvedReleaseVersion -> MVar T.Text -> Maybe SomeException -> IO ()
+handleInstallResult miState ver outputLogVar mError = do
+  installDatas <- atomically $ takeTMVar $ misSdkInstallDatasVar miState
+  let installData = getSdkInstallData ver installDatas
+  case mError of
+    Nothing -> do
+      let homes = sidPendingHomes installData
+          installDatas' = Map.delete ver installDatas
+          disableIde :: SubIdes -> PackageHome -> IO SubIdes
+          disableIde ides home =
+            let ides' = Map.insert home ((lookupSubIde home ides) {ideDataDisabled = IdeDataNotDisabled}) ides
+             in misUnsafeAddNewSubIdeAndSend miState ides' home Nothing
+      atomically $ putTMVar (misSdkInstallDatasVar miState) installDatas'
+      withIDEs_ miState $ \ides -> foldM disableIde ides homes
+    Just err -> do
+      outputLog <- takeMVar outputLogVar
+      let errText = failedInstallIdeDiagnosticMessage ver outputLog err
+      installDatas' <- updateSdkStatus miState installDatas ver LSP.DsError errText (SISFailed outputLog err)
+      sendClient miState $ showMessage LSP.MtError errText
+      atomically $ putTMVar (misSdkInstallDatasVar miState) installDatas'
+
+-- Given a version, and a progress computation, install an sdk (blocking)
+installSdk :: UnresolvedReleaseVersion -> MVar Text -> (Int -> IO ()) -> IO ()
+installSdk unresolvedVersion outputLogVar report = do
+  damlPath <- getDamlPath
+  cachePath <- getCachePath
+  let useCache = mkUseCache cachePath damlPath
+
+  version <- resolveReleaseVersionUnsafe useCache unresolvedVersion
```

Comment:
This may fail to resolve if the cache has not recently been reloaded - we may want to set the `overrideTimeout` field in `useCache` to a lower number than the default.

Reply:
I've set it to 5 minutes, lmk if that seems reasonable